dali/operators/reader/loader/video/video_loader_decoder_gpu.cc

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/loader/video/video_loader_decoder_gpu.h"
+
+namespace dali {
+void VideoSampleGpu::DecodeToOutput(uint8_t *output, cudaStream_t stream) {
+  video_file->SetCudaStream(stream);
+
+  for (int i = 0; i < sequence_len; ++i) {
+    // TODO(awolant): This seek can be optimized - for consecutive frames not needed etc.
+    int frame_id = span->start_ + i * span->stride_;
+    video_file->SeekFrame(frame_id);
+    video_file->ReadNextFrame(output + i * video_file->FrameSize());
+  }
+}
+
+void VideoLoaderDecoderGpu::PrepareEmpty(VideoSampleGpu &sample) {
+  sample = {};
+}
+
+void VideoLoaderDecoderGpu::ReadSample(VideoSampleGpu &sample) {
+  auto &sample_span = sample_spans_[current_index_];
+  auto &video_file = video_files_[sample_span.video_idx_];
+
+  TensorShape<4> sequence_shape = {
+    sequence_len_, video_file.Height(), video_file.Width(), video_file.Channels()};
+
+  sample.data_.Resize(
+    sequence_shape,
+    DALIDataType::DALI_UINT8);
+
+  auto data = sample.data_.mutable_data<uint8_t>();
+
+  // Bind sample to the video and span, so it can be decoded later
+  sample.span = &sample_spans_[current_index_];
+  sample.video_file = &video_files_[sample_span.video_idx_];
+  sample.sequence_len = sequence_len_;
+
+  if (has_labels_) {
+    sample.label_ = labels_[sample_span.video_idx_];
+  }
```

Comment:
I know it's not possible with current code, but what if this "sample" had a label in a previous iteration? we wouldn't be clearing it. How about:
```
sample.label_ = has_labels_ ? labels_[sample_span.video_idx_] : NO_LABEL;
```
being no-label a default value.

Reply:
I added `label_ = -1` in the class, so I don't think we need it here.